pkg/sql/schemachanger/scbuild/builder.go

```diff
@@ -139,6 +152,47 @@ func (b *buildContext) build(
 	return b.outputNodes, nil
 }
 
+func (b *buildContext) addNodeOnce(dir scpb.Target_Direction, elem scpb.Element) {
+	var s scpb.State
+	switch dir {
+	case scpb.Target_ADD:
+		s = scpb.State_ABSENT
+	case scpb.Target_DROP:
+		s = scpb.State_PUBLIC
+	default:
+		panic(errors.Errorf("unknown direction %s", dir))
+	}
+	dropCount := 0
+	newNode := &scpb.Node{
+		Target: scpb.NewTarget(dir, elem),
+		State:  s,
+	}
+	// Check if any existing node matches the new node we are
+	// trying to add.
+	for _, node := range b.outputNodes {
+		if newNode.Element().DescriptorID() == node.Element().DescriptorID() &&
+			reflect.TypeOf(newNode.Element()) == reflect.TypeOf(node.Element()) {
```

Comment:
This screams for some kind of node-equality `Node` or `Element` interface method.